Source/Core/InputCommon/ControllerInterface/Device.h

```diff
@@ -73,37 +117,129 @@ class Device
     // undesirable behavior in our mapping logic.
     virtual bool IsDetectable() const { return true; }
 
+    virtual FocusFlags GetFocusFlags() const { return FocusFlags::Default; }
+
     // Implementations should return a value from 0.0 to 1.0 across their normal range.
     // One input should be provided for each "direction". (e.g. 2 for each axis)
-    // If possible, negative values may be returned in situations where an opposing input is
-    // activated. (e.g. When an underlying axis, X, is currently negative, "Axis X-", will return a
-    // positive value and "Axis X+" may return a negative value.)
+    // If possible, negative values may be returned in situations where an opposing input
+    // is activated. (e.g. When an underlying axis, X, is currently negative, "Axis X-",
+    // will return a positive value and "Axis X+" may return a negative value.)
     // Doing so is solely to allow our input detection logic to better detect false positives.
     // This is necessary when making use of "FullAnalogSurface" as multiple inputs will be seen
-    // increasing from 0.0 to 1.0 as a user tries to map just one. The negative values provide a
-    // view of the underlying axis. (Negative values are clamped off before they reach
+    // increasing from 0.0 to 1.0 as a user tries to map just one. The negative values provide
+    // a view of the underlying axis. (Negative values are clamped off before they reach
     // expression-parser or controller-emu)
     virtual ControlState GetState() const = 0;
 
     Input* ToInput() override { return this; }
-
+	
     // Overridden by CombinedInput,
     // so hotkey logic knows Ctrl, L_Ctrl, and R_Ctrl are the same,
     // and so input detection can return the parent name.
     virtual bool IsChild(const Input*) const { return false; }
+
+protected:
+    InputChannel GetCurrentInputChannel() const;
+    double GetCurrentInputDeltaSeconds() const;
+  };
+
+  //
+  // RelativeInput
+  //
+  // Helper to generate a relative input from an absolute one.
+  // Keeps the last 2 absolute states and returns their difference, divided by the time delta.
+  // It has one state per input channel, as otherwise one SetState() would break
+  // GetState() on the other channels. 
+  // For example the SerialInterface updates inputs at irregular intervals, so to get consistent
+  // results, we need to consider the elapsed time (which also helps keeping in keeping
+  // the values similar between different channels).
+  // You don't have to use this implementation.
+  //
+  template <typename T = ControlState>
+  class RelativeInput : public Input
```

Comment:
I'm not an inputs person so take this whole comment with that in mind but to me this feels overly complex (having seen @jordan-woyak 's [solution](https://github.com/dolphin-emu/dolphin/pull/8747/commits/dbc5d16e3d9b26a265757356bbb23dac8deec42f) ).

In particular, the timing feels like it should be based off the device poll rate, not some general time delta.  After all, we want the relative value between actions.  Plus, we are using seconds precision when it seems we'd want more granularity.

There's a lot of code here so I might just be missing the advantages this provides but wanted to mention for any reviewers that there is another implementation of this feature (my assumption is @jordan-woyak will review so he knows but just in case!).

Reply:
The device poll rate is equal to the input channel delta time as devices. Devices are always updated before reading/caching inputs.

Highly doubt having a double as seconds instead of milliseconds makes any difference (especially given most game engines use float seconds as time delta), but I don't mind changing that.

Reply:
I don't think we should be assuming any sort of usage here.  The devices should be providing the most accurate value they can.  So in this case we should use nanoseconds or whatever precision we have available.

I'm still not sure this is relative input, this feels more like relative rate of change which gives a different value.

Reply:
The delta time I multiply for is fixed for every input channel other that the serial interface, which means it will not have influence on the value (it will not turn it into a speed) other than scaling it.

The serial interface updates at irregular intervals, so if you moved the mouse at a constant speed, the relative input would return a different value for every update, and that is something we don't want (because whatever you actually map it to on GC, it wouldn't need the actual movement, it would need the movement speed). You can see all the thinking behind it in the comment above the time multiplication, the design there is still pending as there isn't a perfect solution unless we add a few settings.

Reply:
Thanks for the explanation @Filoppi !

> Not exactly sure what you mean with nanoseconds but I don't think it applies.

Sorry, yeah, it does not apply.  I had assumed by the name `GetCurrentInputDeltaSeconds()` it was some sort of time diff when in fact it's more of a poll rate for that channel (not actual time delta).

> (because whatever you actually map it to on GC controller, it wouldn't need the actual movement offset, it would need the movement speed)

I still think your value isn't a true relative value so it might be surprising to users.  How do you know users aren't wanting the actual movement offset of the GC controller for their expression?  Feels like we should add some other input function if you want a rate of change (that the user would multiply against).

As a different example, as you might recall I actually want the relative offset of the mouse for FreeLook.  That matches today's master implementation which is what users expect.  Not that I'm that worried about my implementation, I can divide out the scalar or do other workarounds if really necessary.

Anyway, I don't mean to ramble on this. I'm just one person with a view on one tiny part of the code.  I'll let others chime in if they have thoughts.  Great work overall!

Reply:
As I said, yes this would as of now return a rate of change, but given that the time delta is fixed, the rate of change would be the same as the pure change, except scaled, so it does not make a difference anywhere except on the SI.

But yes, theoretically, we should have two mouses axis, one that is the real relative change (for freelook and the wiimote cursor), and one that is the rate of change (to be mapped to analog sticks), but I'm afraid that would be even more confusing. Or sure, we could let users multiply their own relative axis by the input channel time delta, but how would they know they need to do that?

Reply:
I decided to go for the input function to convert a relative axis into a rate of change only when needed. It's a better solution for a lot of reasons and we can explain users they have to do it in one of the help UI tooltips/buttons I added (or on the wiki).
The relative axis code is now cleaned of that massive WIP comment that I needed to figure out what to do with it. Still polishing but using relative axes as input for games has never worked as good as this.

See the "relativeToSpeed(input)" function expression if you want to understand.